From 6c3e1a640ba78b5163a839fa57fd6ae974802cc7 Mon Sep 17 00:00:00 2001
From: wijnand <wikoreman@gmail.com>
Date: Tue, 13 Mar 2018 14:17:00 +0100
Subject: [PATCH] Cosmetics, refactored io queries

---
 avalon/tools/cbsceneinventory/model.py | 12 +++++-------
 1 file changed, 5 insertions(+), 7 deletions(-)

diff --git a/avalon/tools/cbsceneinventory/model.py b/avalon/tools/cbsceneinventory/model.py
index 9e65026..a3a99f5 100644
--- a/avalon/tools/cbsceneinventory/model.py
+++ b/avalon/tools/cbsceneinventory/model.py
@@ -86,11 +86,11 @@ class InventoryModel(TreeModel):
             group the new items with previously existing item groups of the
             same type.
 
-        :param group_items: the items to be processed as returned by `ls()`
-        :type group_items: list
+        Args:
+            items (list): the items to be processed as returned by `ls()`
 
-        :return: root node which has children added based on the data
-        :rtype: node.Node
+        Returns:
+            node.Node: root node which has children added based on the data
         """
 
         # construct the generator results
@@ -110,9 +110,7 @@ class InventoryModel(TreeModel):
             representation = io.find_one({
                 "_id": io.ObjectId(representation_id)
             })
-            version = io.find_one({"_id": representation["parent"]})
-            subset = io.find_one({"_id": version["parent"]})
-            asset = io.find_one({"_id": subset["parent"]})
+            version, subset, asset, _ = io.parenthood(representation)
 
             # Get the primary family
             family = version['data'].get("family", "")
-- 
2.16.2.windows.1

