From 56b5ee676f95688fce8bd4ae732609507019bbbe Mon Sep 17 00:00:00 2001
From: wijnand <wikoreman@gmail.com>
Date: Tue, 13 Mar 2018 16:43:16 +0100
Subject: [PATCH] Restored origin, faster query result

---
 avalon/tools/cbsceneinventory/model.py | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/avalon/tools/cbsceneinventory/model.py b/avalon/tools/cbsceneinventory/model.py
index a3a99f5..84d6698 100644
--- a/avalon/tools/cbsceneinventory/model.py
+++ b/avalon/tools/cbsceneinventory/model.py
@@ -110,7 +110,9 @@ class InventoryModel(TreeModel):
             representation = io.find_one({
                 "_id": io.ObjectId(representation_id)
             })
-            version, subset, asset, _ = io.parenthood(representation)
+            version = io.find_one({"_id": representation["parent"]})
+            subset = io.find_one({"_id": version["parent"]})
+            asset = io.find_one({"_id": subset["parent"]})
 
             # Get the primary family
             family = version['data'].get("family", "")
-- 
2.16.2.windows.1

